lib/tree.rb

```diff
@@ -16,47 +16,126 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log n) if balanced, O(n) if a linked-list type BST
+  # Space Complexity: O(1)
+  def add(key, value = nil)
```

Comment:
👍 However your space complexity is the same as your time complexity because of the call stack (recursion).